spine-contraster.html

```diff
@@ -0,0 +1,152 @@
+        const HEXADECIMAL = 16;
+        const DARK_THEME_MAX_LUMINANCE = 0.179;
+
+        return class SpineContraster extends ParentClass {
+            static get properties() {
+                return {
+                    /**
+                     * The background color value in the format of the CSS rgb or HEX color.
+                     */
+                    bgColor: {
+                        type: String,
+                        observer: '_backgroundObserver'
+                    },
+
+                    /**
+                     * Indicates that the element background is light.
+                     */
+                    lightTheme: {
+                        type: Boolean,
+                        value: false
+                    },
+
+                    /**
+                     * Indicates that the element background is dark.
+                     */
+                    darkTheme: {
+                        type: Boolean,
+                        value: false
+                    }
+                }
+            }
+
+            connectedCallback() {
+                super.connectedCallback();
+                if (!this.bgColor) {
+                    const background = window.getComputedStyle(this)['background-color'];
+                    if (this._isRgb(background)) {
+                        this._calculateTheme(background);
+                    }
+                }
+            }
+
+            /**
+             * For the algorithm details see:
+             *  - https://www.w3.org/TR/WCAG20/#contrast-ratiodef
+             *  - https://www.w3.org/TR/WCAG20/#relativeluminancedef
+             *  - https://en.wikipedia.org/wiki/Luma_(video)#Rec._601_luma_versus_Rec._709_luma_coefficients
+             *
+             * @param background - a color CSS string
+             * @private
+             */
+            _calculateTheme(background) {
+                const isLight = this._isLightTheme(background);
+                this.lightTheme = isLight;
+                this.darkTheme = !isLight;
+            }
+
+            _isLightTheme(background) {
+                const luminance = this._calculateLuminance(background);
+                return luminance > DARK_THEME_MAX_LUMINANCE;
+            }
+
+            _calculateLuminance(color) {
+                const colorRgb = this._toRgb(color);
+                const colors = Object.keys(colorRgb);
+                const tokens = {};
+                for (let i = 0; i < colors.length; i++) {
+                    const color = colors[i];
+                    tokens[color] = colorRgb[color] / 255;
+                    if (tokens[color] <= 0.03928) {
+                        tokens[color] = tokens[color] / 12.92;
+                    } else {
+                        tokens[color] = Math.pow((tokens[color] + 0.055) / 1.055, 2.4);
+                    }
+                }
+                return 0.2126 * tokens.r + 0.7152 * tokens.g + 0.0722 * tokens.b;
```

Comment:
Is it possible to move all figures from `_calculateLuminance` function to constants?